axum-extra/src/middleware/opentelemetry.rs

```diff
@@ -0,0 +1,513 @@
+fn extract_remote_context(headers: &http::HeaderMap) -> opentelemetry::Context {
+    struct HeaderExtractor<'a>(&'a http::HeaderMap);
+
+    impl<'a> opentelemetry::propagation::Extractor for HeaderExtractor<'a> {
+        fn get(&self, key: &str) -> Option<&str> {
+            self.0.get(key).and_then(|value| value.to_str().ok())
+        }
+
+        fn keys(&self) -> Vec<&str> {
+            self.0.keys().map(|value| value.as_str()).collect()
+        }
+    }
+
+    let extractor = HeaderExtractor(headers);
+    opentelemetry::global::get_text_map_propagator(|propagator| propagator.extract(&extractor))
+}
+
+/// Callback that [`Trace`] will call when it receives a request.
+///
+/// [`Trace`]: tower_http::trace::Trace
+#[derive(Clone, Copy, Debug)]
+pub struct OtelOnRequest;
+
+impl<B> OnRequest<B> for OtelOnRequest {
+    #[inline]
+    fn on_request(&mut self, _request: &Request<B>, _span: &Span) {}
+}
+
+/// Callback that [`Trace`] will call when it receives a response.
+///
+/// [`Trace`]: tower_http::trace::Trace
+#[derive(Clone, Copy, Debug)]
+pub struct OtelOnResponse;
+
+impl<B> OnResponse<B> for OtelOnResponse {
+    fn on_response(self, response: &Response<B>, _latency: Duration, span: &Span) {
+        let status = response.status().as_u16().to_string();
+        span.record("http.status_code", &tracing::field::display(status));
+
+        // assume there is no error, if there is `OtelOnFailure` will be called and override this
+        span.record("otel.status_code", &"OK");
+    }
+}
+
+/// Callback that [`Trace`] will call when the response body produces a chunk.
+///
+/// [`Trace`]: tower_http::trace::Trace
+#[derive(Clone, Copy, Debug)]
+pub struct OtelOnBodyChunk;
+
+impl<B> OnBodyChunk<B> for OtelOnBodyChunk {
+    #[inline]
+    fn on_body_chunk(&mut self, _chunk: &B, _latency: Duration, _span: &Span) {}
+}
+
+/// Callback that [`Trace`] will call when a streaming response completes.
+///
+/// [`Trace`]: tower_http::trace::Trace
+#[derive(Clone, Copy, Debug)]
+pub struct OtelOnEos;
+
+impl OnEos for OtelOnEos {
+    #[inline]
+    fn on_eos(self, _trailers: Option<&http::HeaderMap>, _stream_duration: Duration, _span: &Span) {
+    }
+}
+
+/// Callback that [`Trace`] will call when a response or end-of-stream is classified as a failure.
+///
+/// [`Trace`]: tower_http::trace::Trace
+#[derive(Clone, Copy, Debug)]
+pub struct OtelOnFailure;
+
+impl OnFailure<ServerErrorsFailureClass> for OtelOnFailure {
+    fn on_failure(&mut self, failure: ServerErrorsFailureClass, _latency: Duration, span: &Span) {
+        match failure {
+            ServerErrorsFailureClass::StatusCode(status) => {
+                if status.is_server_error() {
+                    span.record("otel.status_code", &"ERROR");
+                }
+            }
+            ServerErrorsFailureClass::Error(_) => {
+                span.record("otel.status_code", &"ERROR");
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
`super::*` imports in tests can "mask" unused imports in the parent module, I would recommend spelling the imports out.

Reply:
I run into that very rarely so I don't think its a big issue. CI should catch that anyway.